enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,585 @@
+	// +kubebuilder:validation:Required
+	MatchPolicy IdentityMatchPolicy `json:"matchPolicy,omitempty"`
+	// exactRepository is required if matchPolicy is set to "ExactRepository".
+	// +optional
+	PolicyMatchExactRepository *ExactRepository `json:"exactRepository,omitempty"`
+	// remapIdentity is required if matchPolicy is set to "RemapIdentity".
+	// +optional
+	PolicyMatchRemapIdentity *RemapIdentity `json:"remapIdentity,omitempty"`
+}
+
+// IdentityMatchPolicy defines the type of matching for setting "matchPolicy".
+// Valid values are "MatchRepoDigestOrExact", "MatchRepository", "ExactRepository", "RemapIdentity". When omitted, the default value is "MatchRepoDigestOrExact".
+type IdentityMatchPolicy string
+
+type PolicyMatchExactRepository struct {
+	// repository is the reference of the image identity to be matched.
+	// +kubebuilder:validation:Required
+	Repository string `json:"repository,omitempty"`
+}
+
+type PolicyMatchRemapIdentity struct {
+	// prefix is the prefix of the image identity to be matched.
+	// +kubebuilder:validation:Required
+	Prefix string `json:"prefix,omitempty"`
+	// signedPrefix is the prefix of the image identity to be matched in the signature.
+	// +kubebuilder:validation:Required
+	SignedPrefix string `json:"signedPrefix,omitempty"`
+}
+
+const (
+	IdentityMatchPolicyMatchRepoDigestOrExact IdentityMatchPolicy = "MatchRepoDigestOrExact"
+	IdentityMatchPolicyMatchRepository        IdentityMatchPolicy = "MatchRepository"
+	IdentityMatchPolicyExactRepository        IdentityMatchPolicy = "ExactRepository"
+	IdentityMatchPolicyRemapIdentity          IdentityMatchPolicy = "RemapIdentity"
+)
+
+// +k8s:deepcopy-gen=true
+type ImagePolicyStatus struct {
+
+	// policyJSON contains the whole policy applied to the kubernetes namespace
+	// which got written to disk. This includes cluster-wide policies
+	// from the `openshift-config` namespace as well.
+	PolicyJSON string `json:"policyJSON,omitempty"`
+
+	// conditions provide details on the status of the gatherer job.
+	Conditions []metav1.Condition `json:"conditions" patchStrategy:"merge" patchMergeKey:"type"`
+}
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Update container runtime config controller to watch ImagePolicy
+
+MCO container runtime config controller ImagePolicy syncHandler updates the signature verification configurations:
+- Retrieves all the ImagePolicy instances on the cluster.
+- Rre-validation: parse the images to validate the policy is not for scopes under OCP product repo. Report the error to machine config logs and the `status` field of that ImagePolicy CR.
+- Validates conflicting data from ImagePolicy instances. Ignore the policies from non global namespace CRs if the sigstore policy for the same image scope has been configured under global(openshift-config) namespace. Log this to the `status` field of both global and non global CRs.
```

Comment:
Why have a special namespace? How would I apply policy to the openshift-config namespace? 

Have you considered a Cluster scoped version of the resource to apply policy to the cluster? A ClusterImagePolicy?

Reply:
It was discussed here: https://redhat-internal.slack.com/archives/CK1AE4ZCK/p1683266421651699 and we want to support the namespace scoped CRD.
cluster admins declare a cluster-scoped default by using 'openshfit-config'.

Reply:
It seems there's motivation to have some support for namespaced, but that conversation also talks about how it's only likely that users are overriding the cluster policy for tests? The conversation reads to me like the cluster scoped policy is much more important.

The precedent for this, in other APIs would be to have a Cluster scoped version of it, then a namespace scoped version of it that handles namespace specifics. I think that would be a good pattern here too.

The cluster admin can handle the cluster scoped version, like they do for ClusterRoles for example, then the namespace admin can handle the namespaced version.

Having a "admin namespace" is complex and non-intuitive. It also opens up a bunch of questions about what happens for pods and other policies in that particular namespace, which could be avoided if we just go with a Cluster scoped CRD for the cluster scoped policy